File: day1/day1.kt
import java.io.File

fun first_digit(line: String): Int {
    var digits =
            mapOf(
                    "one" to 1,
                    "two" to 2,
                    "three" to 3,
                    "four" to 4,
                    "five" to 5,
                    "six" to 6,
                    "seven" to 7,
                    "eight" to 8,
                    "nine" to 9
            )
    for (k in 0..line.length) {
        if (k < line.length && line.get(k).isDigit()) {
            return line.get(k).code - 48
        }
        for (key in digits.keys) {
            if (key in line.substring(0, k)) {
                return digits.getOrDefault(key, 0)
            }
        }
    }
    return 0
}

fun last_digit(line: String): Int {
    var digits =
            mapOf(
                    "one" to 1,
                    "two" to 2,
                    "three" to 3,
                    "four" to 4,
                    "five" to 5,
                    "six" to 6,
                    "seven" to 7,
                    "eight" to 8,
                    "nine" to 9
            )
    for (k in line.length downTo 0) {
        if (k < line.length && line.get(k).isDigit()) {
            return line.get(k).code - 48
        }
        for (key in digits.keys) {
            if (key in line.substring(k, line.length)) {
                return digits.getOrDefault(key, 0)
            }
        }
    }
    return 0
}

fun main() {
    var lines = File("input.txt").readLines()
    var sum = 0
    for (line in lines) {
        sum += (first_digit(line) * 10 + last_digit(line))
    }

    println(sum)
}

File: day1/day1.py
def to_digits(line: str):
    digits = [
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
    ]
    ans = ""
    for i in range(len(line)):
        if line[i].isdigit():
            ans += line[i]
            continue
        for d in range(9):
            if line[i:].startswith(digits[d]):
                ans += str(d + 1)

    return ans


def main():
    with open("input.txt", "r") as f:
        lines = f.readlines()

    sum = 0
    for line in lines:
        digits = to_digits(line)
        sum += int(digits[0]) * 10 + int(digits[-1])

    print(sum)


if __name__ == "__main__":
    main()
